orttraining/orttraining/training_ops/cpu/tensor/gather_grad.cc

```diff
@@ -71,21 +73,24 @@ Status GatherGrad::ComputeImpl(const TensorShape& data_shape, const Tensor& indi
     }
   }
 
-#ifdef USE_OPENMP
-#pragma omp parallel for
-#endif
-  for (int64_t g = 0; g < grad_size; g++) {
+  std::mutex mtx;
+  auto lambda = [&](int64_t g) {
     const int64_t input_block_index = g / output_block_size;
     const int64_t block_offset = g % output_block_size;
     const int64_t indices_index = block_offset / block_size;
     const int64_t offset = block_offset % block_size;
     const Tind idx = indices_data[indices_index];
     const int64_t input_index = input_block_index * input_block_size + idx * block_size + offset;
-#ifdef USE_OPENMP
-#pragma omp atomic
-#endif
+    std::lock_guard<std::mutex> lck(mtx);
```

Comment:
FYI, there are a lot of lock here... Could be a performance bottleneck here... 
Let's add a comment here to point out the potential improvement... 
We don't need to do it now, as we don't use CPU kernel in production training.

Reply:
yea I was thinking the same.

Reply:
And it's often not needed. See: https://papers.nips.cc/paper/4390-hogwild-a-lock-free-approach-to-parallelizing-stochastic-gradient-descent

if it's configurable to have it or not, it will be a little bit better.